browser/ui/ai_chat/print_preview_extractor.cc

```diff
@@ -166,342 +167,383 @@ class PreviewPageTextExtractor {
-  DCHECK(print_preview_ui_id_);
-
-  PrintPreviewDataService::GetInstance()->SetDataEntry(
-      *print_preview_ui_id_, page_index,
-      base::RefCountedSharedMemoryMapping::CreateFromWholeRegion(region));
-}
+  void ClearPreviewUIId() {
+    DCHECK_CURRENTLY_ON(content::BrowserThread::UI);
 
-void PrintPreviewExtractor::OnCompositeToPdfDone(
-    int document_cookie,
-    int32_t request_id,
-    printing::mojom::PrintCompositor::Status status,
-    base::ReadOnlySharedMemoryRegion region) {
-  DVLOG(3) << __func__ << ": id=" << request_id << " , status=" << status;
-  DCHECK_CURRENTLY_ON(content::BrowserThread::UI);
-  DCHECK(print_preview_ui_id_);
-  PrintPreviewDataService::GetInstance()->SetDataEntry(
-      *print_preview_ui_id_, printing::COMPLETE_PREVIEW_DOCUMENT_INDEX,
-      base::RefCountedSharedMemoryMapping::CreateFromWholeRegion(region));
-  OnPreviewReady();
-}
+    if (!print_preview_ui_id_) {
+      return;
+    }
 
-void PrintPreviewExtractor::PreviewCleanup() {
-  if (!print_preview_ui_id_) {
-    return;
-  }
-  PrintPreviewDataService::GetInstance()->RemoveEntry(*print_preview_ui_id_);
-  if (!is_pdf_) {
-    print_render_frame_->OnPrintPreviewDialogClosed();
+    DisconnectPrintPrieviewUI();
+    PrintPreviewDataService::GetInstance()->RemoveEntry(*print_preview_ui_id_);
+    printing::PrintPreviewUI::GetPrintPreviewUIRequestIdMap().erase(
+        *print_preview_ui_id_);
+    printing::PrintPreviewUI::GetPrintPreviewUIIdMap().Remove(
+        *print_preview_ui_id_);
+    print_preview_ui_id_.reset();
   }
-  DisconnectPrintPrieviewUI();
-}
 
-void PrintPreviewExtractor::OnPreviewReady() {
-  scoped_refptr<base::RefCountedMemory> data;
-  CHECK(print_preview_ui_id_);
-  PrintPreviewDataService::GetInstance()->GetDataEntry(
-      *print_preview_ui_id_, printing::COMPLETE_PREVIEW_DOCUMENT_INDEX, &data);
-  if (!data.get()) {
-    DLOG(ERROR) << "no data from preview id: " << *print_preview_ui_id_;
-    return;
-  }
-  auto pdf_region = base::ReadOnlySharedMemoryRegion::Create(data->size());
-  if (!pdf_region.IsValid()) {
-    DLOG(ERROR) << "Failed allocate memory for PDF file";
-    return;
-  }
-  memcpy(pdf_region.mapping.memory(), data->data(), data->size());
-  std::optional<bool> pdf_use_skia_renderer_enabled;
-  auto* prefs = profile_->GetPrefs();
-  if (prefs &&
-      prefs->IsManagedPreference(::prefs::kPdfUseSkiaRendererEnabled)) {
-    pdf_use_skia_renderer_enabled =
-        prefs->GetBoolean(::prefs::kPdfUseSkiaRendererEnabled);
+  void OnPrintPreviewRequest(int request_id) {
+    printing::PrintPreviewUI::GetPrintPreviewUIRequestIdMap()
+        [*print_preview_ui_id_] = request_id;
   }
-  preview_page_text_extractor_ = std::make_unique<PreviewPageTextExtractor>(
-      std::move(pdf_region.region),
-      base::BindOnce(&PrintPreviewExtractor::OnGetOCRResult,
-                     weak_ptr_factory_.GetWeakPtr()),
-      active_chat_tab_helper_->GetMaxPageContentLength(),
```

Comment:
GetMaxPageContentLength was meant for saving unnecessary OCR computing power when we already exceeds the context limit and will be truncated. Maximum 20 pages doesn't prevent this from happening. (It's possible that we reach model context limit first before reaching 20 pages) It is used to prevent edge case like multiple blank pages or small amount of context per page which will take very long to finish OCR when there are significant amount of pages.

Reply:
Is 20 significant? Do you have a different proposal to solve the issue?

As explained over DM a few days ago:
> Also removed max page content consideration since:
> 1) Model can be changed after fetching the content
> 2) Multiple callbacks could occur with different page content limits from different conversations.
> 3) We need to know the (reasonable) total content length to report the percentage of content that Leo has been sent.
> I did consider sending the requested page content maximum length for each GetContent call, but that does not solve all the issues. Since there is a maximum limit of 20 print preview pages, it seems it's ok.

Please remember that multiple conversations can be attached to the same Tab content, and can have different models selected. One other way I could pass a max in is getting the max of all configured models. But that doesn't solve 3). The user would think that Leo is using (almost) 100% the page content when it could be much lower (although still limited to 20 pages).

Reply:
20 is significant when running OCR. If we care so much about reporting percentage of the page content Leo uses, then we can drop max length within PrintPreviewExtrator as you do now. And we should create a follow-up issue for `PrintPreviewExtrator` to report total page count when it hit 20 pages limits so users would know Leo only use parts of the page content.